packages/block-library/src/template-part/edit/index.js

```diff
@@ -27,44 +27,56 @@ export default function TemplatePartEdit( {
 	setAttributes,
 	clientId,
 } ) {
-	const initialPostId = useRef( _postId );
 	const initialSlug = useRef( slug );
 	const initialTheme = useRef( theme );
 
 	// Resolve the post ID if not set, and load its post.
 	const postId = useTemplatePartPost( _postId, slug, theme );
 
-	// Set the post ID, once found, so that edits persist,
-	// but wait until the third inner blocks change,
-	// because the first 2 are just the template part
-	// content loading.
-	const { innerBlocks } = useSelect(
+	// Set the postId attribute if it did not exist,
+	// but wait until the inner blocks have loaded to allow
+	// new edits to trigger this.
+	const { innerBlocks, expectedContent } = useSelect(
 		( select ) => {
 			const { getBlocks } = select( 'core/block-editor' );
+			const entityRecord = select( 'core' ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				postId
+			);
+
 			return {
 				innerBlocks: getBlocks( clientId ),
+				expectedContent: entityRecord?.content.raw,
 			};
 		},
-		[ clientId ]
+		[ clientId, postId ]
 	);
 	const { editEntityRecord } = useDispatch( 'core' );
-	const blockChanges = useRef( 0 );
+	const initialContent = useRef();
 	useEffect( () => {
-		if ( blockChanges.current < 4 ) blockChanges.current++;
+		if ( postId === null || postId === undefined ) {
+			return;
+		}
+
+		if ( ! initialContent.current ) {
+			const innerContent = serialize( innerBlocks );
+			if ( innerContent === expectedContent ) {
+				initialContent.current = innerContent;
+			}
+			return;
+		}
 
 		if (
-			blockChanges.current === 3 &&
-			( initialPostId.current === undefined ||
-				initialPostId.current === null ) &&
-			postId !== undefined &&
-			postId !== null
+			( _postId === null || _postId === undefined ) &&
+			initialContent.current !== serialize( innerBlocks )
 		) {
 			setAttributes( { postId } );
 			editEntityRecord( 'postType', 'wp_template_part', postId, {
 				status: 'publish',
 			} );
 		}
-	}, [ innerBlocks ] );
+	}, [ innerBlocks, expectedContent ] );
```

Comment:
I think this `useEffect` needs some comments, because it's not entirely clear to me how it works.

I guess the idea is that:
- Once we finished loading the template part post...
- If `initialContent` is not set and the block content matches the expected content (aka the template part post content)...
  - We set `initialContent` to the block content.
- Otherwise, if we haven't provided any `postId` to the block (❓), and `initialContent` doesn't match the block content...
  - We "dirty" the template part entity.

This should be a clear demonstration of how clueless I am of what's going on here. 😅

Reply:
Yeah I will try to update comments.  It sounds like you pretty much got the jist of it.

>  We "dirty" the template part entity.

Right!  Technically dirties the entity by making it 'publish' status, and dirties the template it is on by giving 'postId' to the block itself.  And now that the block is given a postId none of the code blocks in these conditions should run anymore.

Reply:
Updated comments.  Let me know if they don't make sense. 😁

Reply:
Yes, thanks! Definitely clearer! 🙇‍♂️